dnn/train_tiny.py

```diff
@@ -0,0 +1,120 @@
+import tensorflow as tf
+import numpy as np
+from tensorflow.keras.callbacks import (
+    ReduceLROnPlateau,
+    EarlyStopping,
+    ModelCheckpoint,
+    TensorBoard
+)
+from dnn.model.yolov3_tiny import YoloV3Tiny, YoloLoss
+from dnn.utils.dataLoader import dataLoader
+from dnn.utils.yolo_encoder import yoloEncoder
+from dnn.parameters import PARAM
+from dnn.utils.mem import tf_set_memory_growth
+
+def main():
+    # Get params
+    size = PARAM['size']
+    anchors = PARAM['yolo_tiny_anchors']
+    anchor_masks = PARAM['yolo_tiny_anchor_masks']
+    num_cls = PARAM['class_number']
+    train_dir = PARAM['train']
+    val_dir = PARAM['val']
+    lr = PARAM['learning_rate']
+    epoch = PARAM['epoch']
+    batch_size = PARAM['batch_size']
+    # Load training and validation data
+    loader = dataLoader(size)
+    train_data = loader(train_dir)
+    val_data = loader(val_dir)
+    # encode data
+    encoder = yoloEncoder(size, anchors, anchor_masks)
+    train_data = train_data.shuffle(buffer_size = 512)
+    train = encoder(train_data)
+
+    # this doesn't work, so still something wrong with dataset
+    test = iter(train)
+    print(next(test))
+
+    val = encoder(val_data)
+    # batch and prefetch
+    train = train.batch(batch_size)
+    train = train.prefetch(buffer_size = tf.data.experimental.AUTOTUNE)
+    val = val.batch(batch_size)
+    val = val.prefetch(buffer_size = tf.data.experimental.AUTOTUNE)
+    # initialize model, loss function and optimizer
+    model = YoloV3Tiny(size, classes=num_cls, training=True)
+    loss = [YoloLoss(anchors[mask], classes=num_cls) 
+            for mask in np.array(anchor_masks, dtype=np.int8)] #cast to int 
+    optimizer = tf.keras.optimizers.Adam(lr = lr)
+    
+    ## Eager training, used for debugging, will remove in the future
+    avg_loss = tf.keras.metrics.Mean('loss', dtype=tf.float32)
+    avg_val_loss = tf.keras.metrics.Mean('val_loss', dtype=tf.float32)
+    for epoch in range(1, epoch + 1):
```

Comment:
This eager training loop is better for debug, but slower than `model.fit`. We should test pieces of the training (network, tf.Dataset, losses) seperately in notebooks. And eventually, put together with `model.fit`.

Another way to be faster is by adding `tf.function` everywhere. But this is quite clunky in my opinion.

Reply:
I was trying to use eager mode to debug first so I commented out ```model.fit``` part.